### PR TITLE
TVS2: Serialization for database-adapters

### PR DESCRIPTION
**Pull request by snazy**
Some `DatabaseAdapter` implementations, for example RocksDB, store certain entities in their serialized form.

Introduces a new artifact `nessie-versioned-persist-serialize` that contains protobuf based serialization.

Objects that are exposed via the `DatabaseAdapter` interface are still based on "immutables" and not protobuf.

---
This change is [Reviewable](https://reviewable.io/reviews/projectnessie/nessie/1765)

### CODE REVIEW

**Review by nastra — APPROVED** [review]

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/DbObjectsSerializers.java [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    }
+    if (sb.length() > 0) {
+      b.addElements(sb.toString());
+      sb.setLength(0);
+    }
+    return b.build();
+  }
+
+  /**
+   * Flatten this {@link Key} into a single string.
```
Comment (rymurr) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/DbObjectsSerializers.java:240]: why deviate from the other parts of the code that serialize . as \00?

**Review by rymurr — CHANGES_REQUESTED** [review]
why straight bytes over protobuf when protobuf is used as our serialization format everywhere else? How do we handle versioning, upgrades, changing of these objects. It seems like while protobuf will create more code for us here it will be more stable long term as this on disk format changes

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/DbObjectsSerializers.java (hunk 1 above)
Comment (snazy) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/DbObjectsSerializers.java:290]: The thought here was that `\000` is maybe not a "nice" character in a `VARCHAR` column

**Comment by snazy** [comment]
PR completely redone w/ protobuf.

**Review by rymurr — DISMISSED** [review]

File: versioned/persist/serialize/src/main/proto/persist.proto [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+message KeyList {
+  repeated KeyWithType keys = 1;
+}
+
+message ContentsId {
+  string id = 1;
+}
+
+message ContentsIdWithType {
+  ContentsId contents_id = 2;
```
Comment (nastra) [inline versioned/persist/serialize/src/main/proto/persist.proto:62]: any particular reason why this doesn't start at `1`? Same for `ContentsIdWithBytes`

**Review by nastra — DISMISSED** [review]

File: versioned/persist/serialize/src/main/proto/persist.proto (hunk 2 above)
Comment (snazy) [reply versioned/persist/serialize/src/main/proto/persist.proto:62]: ehm - no - let me fix that

**Review by nastra — APPROVED** [review]

**Pull request merged**